src/main/java/gumtree/spoon/builder/NodeCreator.java

```diff
@@ -70,16 +71,33 @@ private String getClassName(String simpleName) {
 			ITree variableType = builder.createNode("VARIABLE_TYPE", type.getQualifiedName());
 			variableType.setMetadata(SpoonGumTreeBuilder.SPOON_OBJECT, type);
 			type.putMetadata(SpoonGumTreeBuilder.GUMTREE_NODE, variableType);
-			for (CtTypeReference<?> typeArgument: type.getActualTypeArguments()) {
-				ITree arg = builder.createNode(getClassName(typeArgument.getClass().getSimpleName()), typeArgument.getQualifiedName());
-				arg.setMetadata(SpoonGumTreeBuilder.SPOON_OBJECT, typeArgument);
-				typeArgument.putMetadata(SpoonGumTreeBuilder.GUMTREE_NODE, arg);
-				variableType.addChild(arg);
-			}
+			computeTreeOfTypeReferences(type, variableType);
 			builder.addSiblingNode(variableType);
 		}
 	}
 
+	@Override
+	public void scanCtActualTypeContainer(CtActualTypeContainer reference) {
+		for (CtTypeReference<?> ctTypeArgument: reference.getActualTypeArguments()) {
+			ITree typeArgument = builder.createNode(getClassName(ctTypeArgument.getClass().getSimpleName()), ctTypeArgument.getQualifiedName());
+			typeArgument.setMetadata(SpoonGumTreeBuilder.SPOON_OBJECT, ctTypeArgument);
+			ctTypeArgument.putMetadata(SpoonGumTreeBuilder.GUMTREE_NODE, typeArgument);
+			computeTreeOfTypeReferences(ctTypeArgument, typeArgument);
+			builder.addSiblingNode(typeArgument);
+		}
+	}
+
+	/** Creates a tree of nested type references where each nested type reference is a child of its container. */
+	private void computeTreeOfTypeReferences(CtTypeReference<?> type, ITree parentType) {
+		for (CtTypeReference<?> ctTypeArgument: type.getActualTypeArguments()) {
+			ITree typeArgument = builder.createNode(getClassName(ctTypeArgument.getClass().getSimpleName()), ctTypeArgument.getQualifiedName());
```

Comment:
```suggestion
			ITree typeArgument = builder.createNode(getClassName("TYPE_ARGUMENT"), ctTypeArgument.getQualifiedName());
```

Reply:
I figured that if we pass node type name as the spoon class itself, we get a more accurate diff in my opinion. For example, consider the change `List<A> -> List<String>`. One will get two operations for it, `Delete TypeParameterReference` and `Insert TypeReference`. With your change, there will only be one update operation of `TYPE_ARGUMENT`. Although the semantics of the operations is same, I just felt the former approach gives a more accurate diff of Spoon nodes.

Reply:
> One will get two operations for it, Delete TypeParameterReference and Insert TypeReference

If I understand correctly, the other alternative would only give one UPDATE (ie a shorter edit script)?

Reply:
Yes. Making it `TYPE_ARUGUMENT` would result in a shorted edit script.

Reply:
There is no clear cut decision. It's a design trade-off. 

One way to have both short edit script and accuracy is to put the type name in the node label. Although potentially surprising for users. 

Or to have two children nodes "SUPER_TYPE" and "SUPER_TYPE_TYPE".

Reply:
> Although potentially surprising for users.

I can agree with this as reading "Update TypeReference java.lang.String" is clearer and one doesn't even need to come across "TypeParameterReference".

I will make the changes.

> Or to have two children nodes "SUPER_TYPE" and "SUPER_TYPE_TYPE".

This is not descriptive at all so let's reject this design.

Reply:
Since we have hardcoded typeLabel like "SUPER_TYPE", "THROWN_TYPE", and now "TYPE_ARGUMENT", we can consider refactoring this section of the code a little bit.

`typeLabel` seems more like role in parent so I think we need to rename this parameter as well.